### PR TITLE
PWGHF: Add processData and origin flags in treeCreatorD0ToKPi

### PR DESCRIPTION
**Pull request by atavirag**
I have added flags to distinguish between prompt and non-prompt D0s, and separated between processMC and processData using a PROCESS_SWITCH, so that it can be used on data as well.

### CODE REVIEW

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 1]
```diff
@@ -242,110 +229,53 @@ struct HfTreeCreatorD0ToKPi {
...
-            candidate.originMcRec());
-        }
-      };
-
-      fillTable(0, candidate.isSelD0(), invMassD0ToPiK(candidate), cosThetaStarD0(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
-      fillTable(1, candidate.isSelD0bar(), invMassD0barToKPi(candidate), cosThetaStarD0bar(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
+      auto prong0 = candidate.prong0_as<aod::BigTracksPID>();
+      auto prong1 = candidate.prong0_as<aod::BigTracksPID>();
+      fillTable(candidate, prong0, prong1, 0, candidate.isSelD0(), invMassD0ToPiK(candidate), cosThetaStarD0(candidate), ctD0(candidate), yD0(candidate), eD0(candidate), candidate.flagMcMatchRec(), candidate.originMcRec());
+      fillTable(candidate, prong0, prong1, 1, candidate.isSelD0bar(), invMassD0barToKPi(candidate), cosThetaStarD0bar(candidate), ctD0(candidate), yD0(candidate), eD0(candidate), candidate.flagMcMatchRec(), candidate.originMcRec());
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:278]: The function `fillTable` does not need so many arguments. Providing `candidate, prong0, prong1, 0` is enough. The rest can be retrieved inside the function from `candidate`

Comment (vkucera) [reply PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:278]: Actually, you can keep it this way and avoid calculating `ctD0` and `yD0` twice.

**Comment by atavirag** [comment]
Hi, I am not sure what "clang-format" is asking me to change. I have tried changing the capitalization of acronyms, so that they follow the naming conventions from the documentation, but I still get an error. Do you have any idea what it could be?
Thank you!

**Comment by atavirag** [comment]
Hi @vkucera, would these changes be okay?

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 2]
```diff
@@ -46,27 +46,29 @@ DECLARE_SOA_COLUMN(Eta, eta, float);
...
 DECLARE_SOA_COLUMN(E, e, float);
-DECLARE_SOA_COLUMN(NSigTPCPi0, nsigTPCPi0, float);
-DECLARE_SOA_COLUMN(NSigTPCKa0, nsigTPCKa0, float);
-DECLARE_SOA_COLUMN(NSigTOFPi0, nsigTOFPi0, float);
-DECLARE_SOA_COLUMN(NSigTOFKa0, nsigTOFKa0, float);
-DECLARE_SOA_COLUMN(NSigTPCPi1, nsigTPCPi1, float);
-DECLARE_SOA_COLUMN(NSigTPCKa1, nsigTPCKa1, float);
-DECLARE_SOA_COLUMN(NSigTOFPi1, nsigTOFPi1, float);
-DECLARE_SOA_COLUMN(NSigTOFKa1, nsigTOFKa1, float);
+DECLARE_SOA_COLUMN(NSigTpcPi0, nsigTpcPi0, float);
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:49]:
```suggestion
DECLARE_SOA_COLUMN(NSigTpcPi0, nSigTpcPi0, float);
```
and others

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 3]
```diff
@@ -46,27 +46,29 @@ DECLARE_SOA_COLUMN(Eta, eta, float);
...
 DECLARE_SOA_COLUMN(DecayLengthXYNormalised, decayLengthXYNormalised, float);
-DECLARE_SOA_COLUMN(CPA, cpa, float);
-DECLARE_SOA_COLUMN(CPAXY, cpaXY, float);
+DECLARE_SOA_COLUMN(Cpa, cpa, float);
+DECLARE_SOA_COLUMN(CpaXY, cpaXY, float);
 DECLARE_SOA_COLUMN(Ct, ct, float);
 DECLARE_SOA_COLUMN(ImpactParameterProduct, impactParameterProduct, float);
 DECLARE_SOA_COLUMN(CosThetaStar, cosThetaStar, float);
-DECLARE_SOA_COLUMN(MCflag, mcflag, int8_t);
+DECLARE_SOA_COLUMN(Mcflag, mcflag, int8_t);
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:66]:
```suggestion
DECLARE_SOA_COLUMN(FlagMc, flagMc, int8_t);
```

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 4]
```diff
@@ -154,97 +158,124 @@ struct HfTreeCreatorD0ToKPi {
...
+      collision.posX(),
+      collision.posY(),
+      collision.posZ(),
+      isEventReject,
+      runNumber);
+  }
+
+  template <typename T, typename U, typename V>
+  auto fillTable(const T& candidate, const U& prong0, const V& prong1, int CandFlag, int FunctionSelection, double FunctionInvMass, double FunctionCosThetaStar,
+                 double FunctionCt, double FunctionY, double FunctionE, int8_t FlagMc, int8_t Origin)
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:176]: The arguments are all variables so they should follow the same `camelCase` convention.

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx (hunk 1 above)
Comment (vkucera) [reply PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:278]: Please address the comment and calculate `ctD0` and `yD0` only once.

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 5]
```diff
@@ -154,97 +158,124 @@ struct HfTreeCreatorD0ToKPi {
...
 
-      fillTable(0, candidate.isSelD0(), invMassD0ToPiK(candidate), cosThetaStarD0(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
-      fillTable(1, candidate.isSelD0bar(), invMassD0barToKPi(candidate), cosThetaStarD0bar(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
+  PROCESS_SWITCH(HfTreeCreatorD0ToKPi, processData, "Process data", true);
+
+  void processMc(aod::Collisions const& collisions,
+                 aod::McCollisions const& mccollisions,
+                 soa::Join<aod::HfCand2Prong, aod::HfCand2ProngMcRec, aod::HfSelD0> const& candidates,
+                 soa::Join<aod::McParticles, aod::HfCand2ProngMcGen> const& particles,
+                 aod::BigTracksPID const& tracks)
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:264]: Remove unused variables

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 6]
```diff
@@ -257,10 +288,13 @@ struct HfTreeCreatorD0ToKPi {
...
           particle.phi(),
           RecoDecay::y(array{particle.px(), particle.py(), particle.pz()}, RecoDecay::getMassPDG(particle.pdgCode())),
-          particle.flagMcMatchGen());
+          particle.flagMcMatchGen(),
+          particle.originMcGen());
       }
     }
   }
+
+  PROCESS_SWITCH(HfTreeCreatorD0ToKPi, processMc, "Process Mc", false);
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:297]:
```suggestion
  PROCESS_SWITCH(HfTreeCreatorD0ToKPi, processMc, "Process MC", false);
```

**Review by vkucera — CHANGES_REQUESTED** [review]
Thanks @atavirag , please see further comments.

**Comment by vkucera** [comment]
Add yourself on the list of authors.

**Comment by vkucera** [comment]
The PR title is too vague. Make it more descriptive.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 7]
```diff
@@ -154,97 +158,124 @@ struct HfTreeCreatorD0ToKPi {
...
+        FunctionY,
+        FunctionE,
+        FlagMc,
+        Origin);
+    }
+  }
 
+  void processData(aod::Collisions const& collisions,
+                   soa::Join<aod::HfCand2Prong, aod::HfSelD0> const& candidates,
+                   aod::BigTracksPID const& tracks)
```
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:241]: Here it is `tracks`.

**Comment by atavirag** [comment]
I think it is all changed @vkucera

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 8]
```diff
@@ -154,97 +159,130 @@ struct HfTreeCreatorD0ToKPi {
...
 
-      fillTable(0, candidate.isSelD0(), invMassD0ToPiK(candidate), cosThetaStarD0(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
-      fillTable(1, candidate.isSelD0bar(), invMassD0barToKPi(candidate), cosThetaStarD0bar(candidate), ctD0(candidate), yD0(candidate), eD0(candidate));
+  PROCESS_SWITCH(HfTreeCreatorD0ToKPi, processData, "Process data", true);
+
+  void processMc(aod::Collisions const& collisions,
+                 aod::McCollisions const& mccollisions,
+                 soa::Join<aod::HfCand2Prong, aod::HfCand2ProngMcRec, aod::HfSelD0> const& candidates,
+                 soa::Join<aod::McParticles, aod::HfCand2ProngMcGen> const& particles,
+                 aod::BigTracksPID const&)
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:268]: `mccollisions` doesn't seem used either

File: PWGHF/TableProducer/treeCreatorD0ToKPi.cxx [hunk 9]
```diff
@@ -154,97 +159,130 @@ struct HfTreeCreatorD0ToKPi {
...
+  {
+    // Filling event properties
+    rowCandidateFullEvents.reserve(collisions.size());
+    for (auto& collision : collisions) {
```
Comment (vkucera) [inline PWGHF/TableProducer/treeCreatorD0ToKPi.cxx:272]:
```suggestion
    for (auto const& collision : collisions) {
```
Add missing `const` in all the loops.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by atavirag** [comment]
I just added the changes, sorry for the delay @vkucera

**Review by vkucera — CHANGES_REQUESTED** [review]

**Review by vkucera — APPROVED** [review]
Thanks @atavirag . Please propagate the configuration changes in the validation framework.

**Comment by atavirag** [comment]
I had a look, and I think there are no changes to be made in the validation framework, because the tree creator does not seem to be called.

**Comment by vkucera** [comment]
> I had a look, and I think there are no changes to be made in the validation framework, because the tree creator does not seem to be called.

What do you mean by "does not seem to be called"? The JSON configuration needs to be updates with the process switches.

**Pull request merged**